src/portfft/dispatcher/global_dispatcher.hpp

```diff
@@ -86,70 +87,65 @@ void complex_transpose(const T* a, T* b, IdxGlobal lda, IdxGlobal ldb, IdxGlobal
 }
 
 /**
- * Helper function to determine the increment of twiddle pointer between factors
- * @param level Corresponding implementation for the previous factor
- * @param factor_size length of the factor
- * @return value to increment the pointer by
+ * Copied data from source  to the destination.
+ * @tparam TIn Input Type
+ * @tparam TOut Output Type
+ * @param src sycl::buffer / pointer containing the input data. In the case SPLIT_COMPLEX storage, it contains only the
+ * real part
+ * @param src_imag sycl::buffer / pointer containing the imaginary part of the input in the case where storage is
+ * SPLIT_COMPLEX
+ * @param dst sycl::buffer / pointer containing the output data. In the case SPLIT_COMPLEX storage, it contains only the
+ * real part
+ * @param dst_imag sycl::buffer / pointer containing the imaginary part of the output in the case where storage is
+ * SPLIT_COMPLEX
+ * @param num_elements_to_copy number of elements to copy
+ * @param src_stride distance between two consecutive batches in the input
+ * @param dst_stride disance between two consecutive batches of the output
+ * @param num_copies number of batches to copy
+ * @param input_offset offset applied to the input
+ * @param output_offset offset applied to the output
+ * @param storage complex storage scheme: split_complex / complex_interleaved
+ * @param queue sycl queue associated with the commit
+ * @return
  */
-inline IdxGlobal increment_twiddle_offset(detail::level level, Idx factor_size) {
-  PORTFFT_LOG_FUNCTION_ENTRY();
-  if (level == detail::level::SUBGROUP) {
-    return 2 * factor_size;
-  }
-  if (level == detail::level::WORKGROUP) {
-    Idx n = detail::factorize(factor_size);
-    Idx m = factor_size / n;
-    return 2 * (factor_size + m + n);
+template <typename TIn, typename TOut>
+sycl::event trigger_device_copy(const TIn src, const TIn src_imag, TOut dst, TOut dst_imag,
+                                std::size_t num_elements_to_copy, std::size_t src_stride, std::size_t dst_stride,
+                                std::size_t num_copies, std::size_t input_offset, std::size_t output_offset,
+                                complex_storage storage, sycl::queue& queue) {
+  std::vector<sycl::event> events;
+  auto trigger_device_copy_impl = [&](const TIn& input, TOut& output) {
+    queue.submit([&](sycl::handler& cgh) {
+      auto in_acc_or_usm = get_access(input, cgh);
+      auto out_acc_or_usm = get_access(output, cgh);
+      cgh.host_task([&]() {
+        for (std::size_t i = 0; i < num_copies; i++) {
+          events.push_back(queue.copy(&in_acc_or_usm[0] + i * src_stride + input_offset,
+                                      &out_acc_or_usm[0] + i * dst_stride + output_offset, num_elements_to_copy));
+        }
+      });
```

Comment:
This is undefined behavior for buffers. `queue.copy` is only defined for USM pointers, NOT pointers to buffer memory. However there are `queue.copy` overloads that accept accessors.